// @letta-ai/letta-client/Client.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LettaClient = void 0;
const Client_1 = require("./api/resources/tools/client/Client");
const Client_2 = require("./api/resources/sources/client/Client");
const Client_3 = require("./api/resources/folders/client/Client");
const Client_4 = require("./api/resources/agents/client/Client");
const Client_5 = require("./api/resources/groups/client/Client");
const Client_6 = require("./api/resources/identities/client/Client");
const Client_7 = require("./api/resources/models/client/Client");
const Client_8 = require("./api/resources/embeddingModels/client/Client");
const Client_9 = require("./api/resources/blocks/client/Client");
const Client_10 = require("./api/resources/jobs/client/Client");
const Client_11 = require("./api/resources/health/client/Client");
const Client_12 = require("./api/resources/providers/client/Client");
const Client_13 = require("./api/resources/runs/client/Client");
const Client_14 = require("./api/resources/steps/client/Client");
const Client_15 = require("./api/resources/tags/client/Client");
const Client_16 = require("./api/resources/telemetry/client/Client");
const Client_17 = require("./api/resources/batches/client/Client");
const Client_18 = require("./api/resources/messages/client/Client");
const Client_19 = require("./api/resources/voice/client/Client");
const Client_20 = require("./api/resources/templates/client/Client");
const Client_21 = require("./api/resources/clientSideAccessTokens/client/Client");
const Client_22 = require("./api/resources/projects/client/Client");
class LettaClient {
    constructor(_options = {}) {
        this._options = _options;
    }
    get tools() {
        var _a;
        return ((_a = this._tools) !== null && _a !== void 0 ? _a : (this._tools = new Client_1.Tools(this._options)));
    }
    get sources() {
        var _a;
        return ((_a = this._sources) !== null && _a !== void 0 ? _a : (this._sources = new Client_2.Sources(this._options)));
    }
    get folders() {
        var _a;
        return ((_a = this._folders) !== null && _a !== void 0 ? _a : (this._folders = new Client_3.Folders(this._options)));
    }
    get agents() {
        var _a;
        return ((_a = this._agents) !== null && _a !== void 0 ? _a : (this._agents = new Client_4.Agents(this._options)));
    }
    get groups() {
        var _a;
        return ((_a = this._groups) !== null && _a !== void 0 ? _a : (this._groups = new Client_5.Groups(this._options)));
    }
    get identities() {
        var _a;
        return ((_a = this._identities) !== null && _a !== void 0 ? _a : (this._identities = new Client_6.Identities(this._options)));
    }
    get models() {
        var _a;
        return ((_a = this._models) !== null && _a !== void 0 ? _a : (this._models = new Client_7.Models(this._options)));
    }
    get embeddingModels() {
        var _a;
        return ((_a = this._embeddingModels) !== null && _a !== void 0 ? _a : (this._embeddingModels = new Client_8.EmbeddingModels(this._options)));
    }
    get blocks() {
        var _a;
        return ((_a = this._blocks) !== null && _a !== void 0 ? _a : (this._blocks = new Client_9.Blocks(this._options)));
    }
    get jobs() {
        var _a;
        return ((_a = this._jobs) !== null && _a !== void 0 ? _a : (this._jobs = new Client_10.Jobs(this._options)));
    }
    get health() {
        var _a;
        return ((_a = this._health) !== null && _a !== void 0 ? _a : (this._health = new Client_11.Health(this._options)));
    }
    get providers() {
        var _a;
        return ((_a = this._providers) !== null && _a !== void 0 ? _a : (this._providers = new Client_12.Providers(this._options)));
    }
    get runs() {
        var _a;
        return ((_a = this._runs) !== null && _a !== void 0 ? _a : (this._runs = new Client_13.Runs(this._options)));
    }
    get steps() {
        var _a;
        return ((_a = this._steps) !== null && _a !== void 0 ? _a : (this._steps = new Client_14.Steps(this._options)));
    }
    get tags() {
        var _a;
        return ((_a = this._tags) !== null && _a !== void 0 ? _a : (this._tags = new Client_15.Tags(this._options)));
    }
    get telemetry() {
        var _a;
        return ((_a = this._telemetry) !== null && _a !== void 0 ? _a : (this._telemetry = new Client_16.Telemetry(this._options)));
    }
    get batches() {
        var _a;
        return ((_a = this._batches) !== null && _a !== void 0 ? _a : (this._batches = new Client_17.Batches(this._options)));
    }
    get messages() {
        var _a;
        return ((_a = this._messages) !== null && _a !== void 0 ? _a : (this._messages = new Client_18.Messages(this._options)));
    }
    get voice() {
        var _a;
        return ((_a = this._voice) !== null && _a !== void 0 ? _a : (this._voice = new Client_19.Voice(this._options)));
    }
    get templates() {
        var _a;
        return ((_a = this._templates) !== null && _a !== void 0 ? _a : (this._templates = new Client_20.Templates(this._options)));
    }
    get clientSideAccessTokens() {
        var _a;
        return ((_a = this._clientSideAccessTokens) !== null && _a !== void 0 ? _a : (this._clientSideAccessTokens = new Client_21.ClientSideAccessTokens(this._options)));
    }
    get projects() {
        var _a;
        return ((_a = this._projects) !== null && _a !== void 0 ? _a : (this._projects = new Client_22.Projects(this._options)));
    }
}
exports.LettaClient = LettaClient;
